Update caching and performance docs (#5784)
AzureAD/microsoft-authentication-library-for-js: lib/msal-node/docs/caching.md

Reviewer: When working through MSAL it was not initially clear to me that I would need to use the integrated cache as refreshToken is not exposed. I think in this document it would be really helpful to document this at the top or even in the README.
1. Explicitly state what the MSAL cache schema format is. I assume it's similar to the exposed token information, but also includes the refreshToken.
2. Explain within this document that the refreshToken is intentionally not included in msal responses and the reasoning behind it.

Author: Sounds good, added a remark about refresh token usage. I think we intentionally didn't expose the cache schema itself, as its an implementation detail and might be subject to change in future. In general, developers shouldn't need to take a dependency on the cache schema, and if the available APIs don't cover every need there, that's likely a problem or bug (fyi, cache schema is [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/cache.json))

Reviewer 2: I also think the reasoning behind `RT not being shared to the users` and `cache schema` are two distinct things and are orthogonal. We can take the RT q in an FAQ and may be link here if that is more appropriate.

In production, you would most likely want to serialize and persist the token cache. Depending on the type of application, you can:

* Desktop apps, console apps (public clients apps (PCA)):
  * Use [MSAL Node Extensions](../../../extensions/msal-node-extensions/README.md), which provides persistence and encryption at rest solutions on Windows, Linux and Mac OS
* Web apps, web APIs, daemon apps (confidential client apps (CCA)):
  * MSAL's in-memory token cache does not scale for production. Use the [distributed token caching](#performance-and-security) pattern to persist the cache in your choice of storage environment (Redis, MongoDB, SQL databases etc. -keep in mind that you can use these in tandem *e.g.* a Redis-like memory cache as a first layer of persistence, and a SQL database as a second, more stable persistence layer)

## In-memory cache

> By default, MSAL has an in-memory cache. The in-memory cache is the single source of truth for all MSAL activities involving tokens and accounts. The lifetime of in-memory cache is the same as the MSAL instance. If the MSAL instance restarts, the cache is erased when the process lifecycle finishes. We recommend persisting the cache with encryption for all production applications both for security and desired cache longevity. If you choose not to persist the cache, the [TokenCache](https://azuread.github.io/microsoft-authentication-library-for-js/ref/classes/_azure_msal_node.tokencache.html) interface is still available to access the cached entities.

Reviewer 3: ```suggestion
MSAL maintains an in-memory cache. The in-memory cache is the single source of truth for all MSAL activities involving tokens and accounts. The lifetime of in-memory cache is the same as the MSAL application object. If the process using MSAL restarts, the cache is erased when the process lifecycle finishes. This affects user scenarios as users will have to re-authenticate, and although the browser will do most some of the work, the end user experience suffers. Service to service scenarios suffer because getting a token from AAD involves HTTP requests, and is much slower than getting a token from a cache.

 We recommend persisting the cache with encryption for all production applications both for security and desired cache longevity. If you choose not to persist the cache, the [TokenCache](https://azuread.github.io/microsoft-authentication-library-for-js/ref/classes/_azure_msal_node.tokencache.html) interface is still available to access the cached entities.
```

In production, you would most likely want to serialize and persist the token cache. Depending on the type of application, you can:

* Desktop apps, console apps (public clients apps (PCA)):
  * Use [MSAL Node Extensions](../../../extensions/msal-node-extensions/README.md), which provides persistence and encryption at rest solutions on Windows, Linux and Mac OS
* Web apps, web APIs, daemon apps (confidential client apps (CCA)):
  * MSAL's in-memory token cache does not scale for production. Use the [distributed token caching](#performance-and-security) pattern to persist the cache in your choice of storage environment (Redis, MongoDB, SQL databases etc. -keep in mind that you can use these in tandem *e.g.* a Redis-like memory cache as a first layer of persistence, and a SQL database as a second, more stable persistence layer)

## In-memory cache

> MSAL maintains an in-memory cache. The in-memory cache is representative of the application cache state. The lifetime of in-memory cache is the same as the MSAL application object. If the process using MSAL restarts, the cache is erased when the process lifecycle finishes. This affects user scenarios as users will have to re-authenticate, and although the browser will do most some of the work, the end user experience suffers. Service to service scenarios also suffer because getting a token from AAD involves HTTP requests, and is much slower than getting a token from a cache.

Reviewer: ```suggestion
MSAL maintains an in-memory cache. The in-memory cache is representative of the application cache state. The lifetime of in-memory cache is the same as the MSAL application object. If the process using MSAL restarts, the cache is erased when the process lifecycle finishes. This affects user scenarios as users will have to re-authenticate, and although the browser will do most of the work, the end user experience suffers. Service to service scenarios also suffer because getting a token from AAD involves HTTP requests, and is much slower than getting a token from a cache.
```

Reviewer: I'm by no means a wordsmith but the last section is a bit confusing. Might be worth condensing to clarify what is relevant for browser or backend service
```suggestion
MSAL maintains an in-memory cache. The in-memory cache is representative of the application cache state. The lifetime of in-memory cache is the same as the MSAL application object. If the process using MSAL restarts, the cache is erased when the process lifecycle finishes. This affects user scenarios as users will have to re-authenticate. This means a worse user experience in browser or an expensive HTTP request in service to service scenarios. 
```

Author: that's fair. I tried to reword this, hope its a bit more clear now

> MSAL maintains an in-memory cache. The in-memory cache is the single source of truth for all MSAL activities involving tokens and accounts. The lifetime of in-memory cache is the same as the MSAL application object. If the process using MSAL restarts, the cache is erased when the process lifecycle finishes. This affects user scenarios as users will have to re-authenticate, and although the browser will do most some of the work, the end user experience suffers. Service to service scenarios also suffer because getting a token from AAD involves HTTP requests, and is much slower than getting a token from a cache.

Reviewer 2: Is it really "single source of truth"? It is representative of the application cache state isn't it?

Author: Sorry for the late response. What I'm after here is to prevent any confusion about which cache the acquireToken* APIs primarily rely on. I think "representative of the application cache state" sounds more accurate though -will revise

> Note that the in-memory cache is not scalable for server-side applications and performance will degrade after holding a few 100 tokens in cache. For web app and web API scenarios, this approximates to serving a few 100 users. For daemon app scenarios using client credentials grant to call other apps, this means a few 100 tenants. See [performance](#performance-and-security) below for more.

Reviewer: ```suggestion
Note because this cache strategy is stored in-memory, it is not scalable for server-side applications and performance will degrade after holding a few 100 tokens in cache. For web app and web API scenarios, this approximates to serving a few 100 users. For daemon app scenarios using client credentials grant to call other apps, this means a few 100 tenants. See [performance](#performance-and-security) below for more.
```

> > :warning: We recommend persisting the cache with encryption for all production applications both for security and desired cache longevity. If you choose not to persist the cache, the [TokenCache](https://azuread.github.io/microsoft-authentication-library-for-js/ref/classes/_azure_msal_node.tokencache.html) interface is still available to access the cached entities.

Reviewer 2: `encryption` is lost in the message. Can we make sure it is highlighted and both `persistence` and `encryption` are handled in the following sentence?

Author: done!

## Persistent cache

> MSAL Node fires events when the in-memory cache is accessed and apps can choose whether to serialize or deserialize the cache (see: [TokenCacheContext](https://azuread.github.io/microsoft-authentication-library-for-js/ref/classes/_azure_msal_common.tokencachecontext.html)) to persistence (e.g. a file, a SQL database and etc.). This constitutes two actions:

Reviewer 2: Do we need to mention serialization? It is just a technique to ensure cache integrity. We can just say `persist` here. If you believe you need the advanced users to know the details for performance reasons, I prefer to separate the description and expand on why we settled on a `json` format to persist cache and we apply serialization and de-serialization to achieve the same.

Reviewer 2: I see why you may have done this, as the API name consists of it. However, the above still applies.

Author: Sure thing, used save/load instead to be agnostic about this -there's enough hints elsewhere about using the serialization technique

}
```

A basic implementation of `ICachePlugin` interface might look as follows:

```typescript
class MyCachePlugin implements ICachePlugin {
    private client: ICacheClient;

    constructor(client: ICacheClient) {
        this.client = client; // client object to access the persistent cache
    }

    public async beforeCacheAccess(cacheContext: TokenCacheContext): Promise<void> {

>         const cacheData = await this.client.get(); // get the cache from persistence

Reviewer: Shouldn't there be a key passed in get/set for client?

Author: the basic use cases (for instance, in a desktop or console app), you may not necessarily need a key, but I added a remark to read further for server side scenarios (and I think link to the samples cover this)

    }

    public async afterCacheAccess(cacheContext: TokenCacheContext): Promise<void> {
        if (cacheContext.cacheHasChanged) {
            await this.client.set(cacheContext.tokenCache.serialize()); // deserialize in-memory cache to persistence
        }
    }
}
```

* If you are developing a public client app, [MSAL Node Extensions](../../../extensions/msal-node-extensions/README.md) handles this for you.
* If you are developing a confidential client app, you should persist the cache via a separate service, since a single, *per-server* cache instance isn't suitable for a cloud environment with many servers and app instances.

> :warning: We strongly recommend to encrypt the token cache when persisting it on disk. For public client apps, this is offered out-of-box with [MSAL Node Extensions](../../../extensions/msal-node-extensions/README.md). For confidential clients however, you are responsible for devising an appropriate encryption solution.

## Performance and security

On public client apps, [MSAL Node Extensions](../../../extensions/msal-node-extensions/README.md) ensures performance and security for you.

> On confidential client apps that handle users (web apps that sign in users and call web APIs, and web APIs calling downstream web APIs), there can be many users and the users are processed in parallel. Our recommendation is to serialize one cache blob (see [CacheRecord](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-common/src/cache/entities/CacheRecord.ts)) per user. Use a key for partitioning the cache (*i.e.* **partition key**), such as:

Reviewer 2: ```suggestion
On confidential client apps that handle users (web apps that sign in users and call web APIs, and web APIs calling downstream web APIs), there can be many users active concurrently for a given application. Our recommendation is to serialize one cache blob (see [CacheRecord](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-common/src/cache/entities/CacheRecord.ts)) per user. This would help with scaling the cache across a distributed system. Use a key for partitioning the cache (*i.e.* **partition key**), such as:
```
